src/compiler/transformers/es2015.ts

```diff
@@ -3778,16 +3778,42 @@ namespace ts {
          * @param node An Identifier node.
          */
         function substituteExpressionIdentifier(node: Identifier): Identifier {
-            if (enabledSubstitutions & ES2015SubstitutionFlags.BlockScopedBindings) {
+            if (enabledSubstitutions & ES2015SubstitutionFlags.BlockScopedBindings && !isInternalName(node)) {
                 const declaration = resolver.getReferencedDeclarationWithCollidingName(node);
-                if (declaration) {
+                if (declaration && !(isClassLike(declaration) && isPartOfClassBody(declaration, node))) {
                     return setTextRange(getGeneratedNameForNode(declaration.name), node);
                 }
             }
 
             return node;
         }
 
+        function isPartOfClassBody(declaration: ClassLikeDeclaration, node: Identifier) {
+            let currentNode = getParseTreeNode(node);
+            if (!currentNode || currentNode === declaration || currentNode.end <= declaration.pos || currentNode.pos >= declaration.end) {
+                // if the node has no correlation to a parse tree node, its definitely not
+                // part of the body.
+                // if the node is outside of the document range of the declaration, its
+                // definitely not part of the body.
+                return false;
+            }
+            const blockScope = getEnclosingBlockScopeContainer(declaration);
+            while (currentNode) {

```

Comment:
`findAncestor` doesn't do much to improve this. It's the same number of lines, introduces a new environment record for the closure each time it's called, and doesn't improve readability of the function. There's no apparent performance difference between the two, as far as I can tell.

```ts
        function isPartOfClassBody(declaration: ClassLikeDeclaration, node: Identifier) {
            node = getParseTreeNode(node, isIdentifier);
            if (!node || node.end <= declaration.pos || node.pos >= declaration.end) {
                // if the node has no correlation to a parse tree node, its definitely not
                // part of the body.
                // if the node is outside of the document range of the declaration, its
                // definitely not part of the body.
                return false;
            }
            const blockScope = getEnclosingBlockScopeContainer(declaration);
            return !!findAncestor(node, current => {
                if (current === blockScope || current === declaration) {
                    return "quit";
                }
                if (isClassElement(current) && current.parent === declaration) {
                    // we are in the class body, but we treat static fields as outside of the class body
                    return current.kind !== SyntaxKind.PropertyDeclaration
                        || (getModifierFlags(current) & ModifierFlags.Static) === 0 ? true : "quit";
                }
                return false;
            });
        }
```